Assortativity extended

vboussange opened the pull request:
Hi there, I propose an extension of method `assortativity` to account for scalar properties of nodes:

This is also done in [networkX (python)](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.assortativity.attribute_assortativity_coefficient.html#networkx.algorithms.assortativity.attribute_assortativity_coefficient).

"""
    assortativity(g,labels)
Similar to `assortativity(g)` except that Pearson correlation is calculated
from the correlation between some values associated to each node and stored in
`labels`.

# Arguments
- `labels` is a dictionary that associates to each vertex index a scalar value

# Examples
```jldoctest
julia> using LightGraphs

julia> labels = Dict(collect(1:4) .=> [-1., -1., 1., 1.])

julia> assortativity(star_graph(4),labels)
-0.5
```
"""

sbromberger commented:
Thanks for these PRs, @vboussange ! In order for us to merge them, they will need comprehensive tests. 

Also, a formal review is forthcoming, but why are you using a `Dict` for vertex indices instead of a `Vector{T}`? Vectors are much more efficient.

vboussange commented:
> Thanks for these PRs, @vboussange ! In order for us to merge them, they will need comprehensive tests.
> 
> Also, a formal review is forthcoming, but why are you using a `Dict` for vertex indices instead of a `Vector{T}`? Vectors are much more efficient.

@sbromberger Thanks for your comments.  I was using `Dict` because of personal historical contingencies. You are right - `Vector` makes much more sense. I have made a new commit with the updated function, added some tests, and modified the documentation accordingly.

sbromberger commented:
sorry for the delay in review - I should be able to get to this this week.

sbromberger commented on src/community/assortativity.jl:
```diff
@@ -1,21 +1,35 @@
 """
     assortativity(g)
+    assortativity(g,attributes)
```
space between comma and "attributes"

sbromberger commented on src/community/assortativity.jl:
```diff
@@ -34,18 +48,33 @@ function assortativity(g::AbstractGraph{T}) where T
     return assortativity_coefficient(g, sjk, sj, sk, sjs, sks, nue)
 end
 
+function assortativity(g::AbstractGraph{T},attributes::Vector{N}) where {T,N<:Number}
```
spaces between arguments, please

sbromberger commented on src/community/assortativity.jl:
```diff
@@ -34,18 +48,33 @@ function assortativity(g::AbstractGraph{T}) where T
+        sjk += j*k
+        sj  += j
+        sk  += k
+        sjs += j^2
```
under what circumstances will `sjs` and `sks` overflow? If `N` is `Int8`, for example, and one of the elements in attributes is `7`, then you've got a potential problem.

sbromberger commented on test/community/assortativity.jl:
```diff
@@ -20,4 +20,28 @@ using Random, Statistics
+    # between some scalar attributes characterising each vertex
+    @testset "Small graphs with attributes" for n = 5:2:11
+        g = grid([1,n])
+        attributes = [(-1.)^v for v in vertices(g)]
```
it's not a good idea to assume `vertices` is an integer that you can use (it is for now, but not a great idea to rely on this behavior). Use `1:nv(g)` for times when you want a numeric value.

sbromberger commented on test/community/assortativity.jl:
```diff
@@ -20,4 +20,28 @@ using Random, Statistics
+    # Test definition of assortativity as Pearson correlation coefficient
+    # between some scalar attributes characterising each vertex
+    @testset "Small graphs with attributes" for n = 5:2:11
+        g = grid([1,n])
```
check spacing between arguments. `grid([1, n])` is more legible.

sbromberger commented:
Best of intentions, etc. Sorry for the delay. Reviewing now.

sbromberger commented on src/community/assortativity.jl:
```diff
@@ -34,18 +48,34 @@ function assortativity(g::AbstractGraph{T}) where T
     return assortativity_coefficient(g, sjk, sj, sk, sjs, sks, nue)
 end
 
+function assortativity(g::AbstractGraph{T}, attributes::Vector{N}) where {T,N<:Number}
```
`T` needs to be `<: Integer`. Can `N` really be `<: Number`? Does this work with complex?

vboussange replied:
Thanks for your comment. `T` is now as requested `<: Integer`. You are right, `N` can not be `Complex`, I thus modified it as `<:Real`.
